src/linker/Linker.Dataflow/XmlFlowAnnotationSource.cs

```diff
@@ -28,76 +26,109 @@ public XmlFlowAnnotationSource (LinkContext context)
 			_context = context;
 		}
 
-		public DynamicallyAccessedMemberTypes GetFieldAnnotation (FieldDefinition field)
+		public IEnumerable<CustomAttribute> GetCustomAttributes (ICustomAttributeProvider provider)
 		{
-			return _fields.TryGetValue (field, out var ann) ? ann : DynamicallyAccessedMemberTypes.None;
+			return _attributes.TryGetValue (provider, out var ann) ? ann : null;
 		}
 
-		public DynamicallyAccessedMemberTypes GetParameterAnnotation (MethodDefinition method, int index)
+		public bool HasCustomAttributes (ICustomAttributeProvider provider)
 		{
-			DynamicallyAccessedMemberTypes parameterAnnotation = DynamicallyAccessedMemberTypes.None;
-
-			if (_methods.TryGetValue (method, out var ann) && ann.ParameterAnnotations != null) {
-				string paramName = method.Parameters[index].Name;
-
-				bool firstAppearance = true;
-				foreach (var (ParamName, Annotation) in ann.ParameterAnnotations) {
-					if (ParamName == paramName && firstAppearance) {
-						firstAppearance = false;
-						parameterAnnotation = Annotation;
-					} else if (ParamName == paramName && !firstAppearance) {
-						_context.LogMessage (MessageContainer.CreateWarningMessage ($"There are duplicate parameter names for '{paramName}' inside '{method.Name}' in '{_xmlDocumentLocation}'", 2024));
-					}
-				}
-			}
-
-			return parameterAnnotation;
+			return _attributes.ContainsKey (provider);
 		}
 
-		public DynamicallyAccessedMemberTypes GetPropertyAnnotation (PropertyDefinition property)
+		IEnumerable<CustomAttribute> ProcessAttributes (XPathNavigator nav)
 		{
-			return _properties.TryGetValue (property, out var ann) ? ann : DynamicallyAccessedMemberTypes.None;
-		}
+			XPathNodeIterator iterator = nav.SelectChildren ("attribute", string.Empty);
+			var attributes = new List<CustomAttribute> ();
+			while (iterator.MoveNext ()) {
+				AssemblyDefinition assembly;
+				TypeDefinition attributeType;
 
-		public DynamicallyAccessedMemberTypes GetReturnParameterAnnotation (MethodDefinition method)
-		{
-			return _methods.TryGetValue (method, out var ann) ? ann.ReturnAnnotation : DynamicallyAccessedMemberTypes.None;
-		}
+				string attributeFullName = GetFullName (iterator.Current);
+				if (attributeFullName == String.Empty) {
+					_context.LogMessage (MessageContainer.CreateWarningMessage ($"Attribute element does not contain attribute 'fullname'", 2029));
+					continue;
+				}
+				string concatAttribute;
+				concatAttribute = attributeFullName.EndsWith ("Attribute") ? attributeFullName : String.Concat (attributeFullName, "Attribute");
```

Comment:
Why is the rationale behind this logic?

Reply:
I was thinking about if someone would use the abbreviate names, in those cases append the Attribute at the end. But yeah, it may cause some confusion. Deleted it and now the user will need to write the actual attribute name.